### PR TITLE
feat: validate signature by user

### PR DESCRIPTION
**Pull request by 0xAaCE**
# feat: validate signature by user

## Changelog

- When we have to sign a risky transaction we gonna ask for user validation

### Type of changes included:

- [ ] Components
- [x] Business Logic
- [ ] Bug fixes
- [ ] Styling
- [ ] Code Refactor

### Clubhouse Tickets Related:

- https://app.shortcut.com/terminalsystems/story/24132/v4-ic-provider-refactor-whitelist-flow

### Notes:

TODO: design and implement a nicer modal for showing information of transaction

### Tested on:

- [ ] Chrome
- [ ] Firefox
- [ ] Safari

### CODE REVIEW

File: source/Background/Controller.js [hunk 1]
```diff
@@ -380,30 +381,86 @@ backgroundController.exposeController(
...
           callback(ERRORS.CONNECTION_ERROR, null);
           return;
         }
-        const parsedPayload = payload instanceof Buffer
-          ? payload
-          : Buffer.from(Object.values(payload));
-        const signed = await keyring.sign(parsedPayload);
-        callback(null, [...new Uint8Array(signed)]);
+
+        if (requestType !== 'read_state' && canisterId && !(canisterId in Object.values(app.whitelist))) {
```
Comment (tomiir) [inline source/Background/Controller.js:400]: what's the whitelst format? thought it was an array

File: source/Background/Controller.js [hunk 2]
```diff
@@ -380,30 +381,86 @@ backgroundController.exposeController(
...
+          callback(ERRORS.CANISTER_NOT_WHITLESTED_ERROR(canisterId), null);
+          return;
+        }
+
+        const canisterInfo = app.whitelist[canisterId];
+
+        if ((requestInfo.requestType === 'call' /*&& canisterInfo.category in PROTECTED_CATEGORIES*/) || requestInfo.manual) {
```
[the remark this answers is not in the document]
Comment (tomiir) [reply source/Background/Controller.js:407]: is there any other type other than call and read_state?

File: source/Background/Controller.js [hunk 3]
```diff
@@ -380,30 +381,86 @@ backgroundController.exposeController(
...
+            url,
+            type: 'popup',
+            width: SIZES.width,
+            height,
+          });
+        } else {
+          const parsedPayload = new Uint8Array(Object.values(payload));
+
+          const signed = await keyring.sign(parsedPayload.buffer);
+          callback(null, [...new Uint8Array(signed)]);
```
Comment (tomiir) [inline source/Background/Controller.js:432]: wont this deserialize it again? Like why construct the uint8array here?

File: source/Pages/Notification/components/AllowAgent/index.jsx [hunk 4]
```diff
@@ -56,9 +56,11 @@ const AllowAgent = ({
   const [expand, setExpand] = useState(false);
 
   const handleAllowAgent = async (status) => {
+    const whitelist = {};
+    canistersInfo.forEach(canisterInfo => whitelist[canisterInfo.id] = { canisterInfo })
```
[the remark this answers is not in the document]
Comment (maksym-boytsov) [reply source/Pages/Notification/components/AllowAgent/index.jsx:60]: It's fine, like we could do `reduce` for sure to make it look better

File: source/Background/Controller.js (hunk 1 above)
Comment (rocky-fleek) [reply source/Background/Controller.js:400]: with what format?

Comment (aleacevedo) [reply source/Background/Controller.js:400]:
```
{
  [canisterId]: { canisterDabInfo }
}
```

**Pull request merged**